# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [11]:
def add_two(x):
    return x*2

#then we can test it:
print(add_two(4))
#This is great for continious, but to only do it once, we used lambda
a_var = 4
print((lambda x: x+2)(a_var))

8
6


In terms of time and space complexity, they are the same in performance as a normal function.

Lambda should be used when we want to do a single function

#### Saving to a Variable

In [ ]:
# we can save a lambda function to a variable for reusability
# do not recommend, because a function can do the same thing
f_test = lambda x: x+2
f_test(4)

In [ ]:
f_fest(78)

#### Multiple Inputs

In [1]:
# No variable
print((lambda x,y,z,: x  * y * z)(3,5,8))

# if we wanted to save the lambda function as a variable, we could as well with multiple inputs:
x_test = lambda x,y,z,: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [ ]:
def multiply(f, num):
    """
    f expects to see a function(lambda_function)
    num expects a single number
    """
    return f(num)
multiply(lambda x: x*x, 4)

# not only can we pass lambda functions into another function
def squared_nums(num):
    return num ** 2
multiply(squared_nums, 4)


#### Returning a Lambda from a Function

In [7]:
def multiply_test(num):
    return num * 4

# Let's try function inside another function
def return_func():
    test = 4
    def multiply(num):
        return num*2
    return multiply

f_return = return_func()
print(f_return(4))

# what if i wanted to return a lambda function from a regular function
def return_lamb(b,c):
    return lambda x,a: x+a+b+c

r_lamb = return_lamb(4,5) # this is the b,c
print(r_lamb)
print(r_lamb(5,5)) # this is mirroring the x,a, this combine bc(x,a)

8
<function return_lamb.<locals>.<lambda> at 0x000001D829905000>
19


#### If Statements within Lambdas

In [9]:
# lambdas are set up to be able to take in conditional statements
# lambda x: True if condition happens, else False
# lambda x: do something if conditional happens else do something else
f_conditional = lambda num: num * 2 if num > 10 else num + 2
print(f_conditional(8))
print(f_conditional(20))

10
40


In [10]:
lst = [1,2,3,4,5,6,7,8,9,10,11,12,13]
[f_conditional(num) for num in lst]

[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 22, 24, 26]

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
f = lambda x: x ** 3
print(f(2))

8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [17]:
numbers = [4,11,20,20,3,15,20,11]

# normally, in order ot use a function with this, we would iterate through this list
def squared_normal(val):
    return val **2
lst = []
for num in numbers:
    lst.append(squared_normal(num))
print(lst)


[16, 121, 400, 400, 9, 225, 400, 121]


In [2]:
#
"""
map(function, iterable)
An iterable is something we can loop over; such as a list, dict, tuple, sat, or a string.
Normally, you would use a map with a predefined function, but we can use with a lambda
"""
%time
def squared_map(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2 **2
    else:
        return num, num2

numbers = [4,11,20,20,3,15,20,11]
more_nums = [3,15,20,11,4,11,20,20]

# squared(numbers, more_nums) # would not work because num and num2 are expecting int not lists
# You would need to zip() function which brings all integers together
# 
for val1, val2, in zip(numbers, more_nums):
    print(squared_map(val1, val2))
'''
when we use map against two or more iterable objects, we get returned a 
list of tuples. each value in the tuple represents a value in the 
corresponding input list
'''
squared_numbers_mapped = list(map(squared_map, numbers, more_nums))
print(squared_numbers_mapped)

# when we use map against a single iterable object, we will return a list of values
#where each value corresponds to a value in the input object
another_squared_nums_mapped = list(map(squared_normal, numbers))
print(another_squared_nums_mapped)

CPU times: total: 0 ns
Wall time: 0 ns
(16, 9)
(11, 15)
(20, 20)
(20, 11)
(9, 16)
(15, 11)
(20, 20)
(11, 20)
[(16, 9), (11, 15), (20, 20), (20, 11), (9, 16), (15, 11), (20, 20), (11, 20)]


NameError: name 'squared_normal' is not defined

#### Using Lambda's with Map

In [ ]:
"""
Instead of having to build out a function that we plan to use one time
We can use a lambda inside the map function. Allowing us to do everything
in a single line of code

Syntax for this:
map(lambda x: do something, iterable)
Once more, using lambda in map almost always takes place in a single line of code
"""

#remember that we need to use list function or it'll give you a hex code
list(map(lambda x,y: (x**2, y**2) if x < 10 and y < 10 else (x,y), numbers, more_nums) )

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [18]:
numbers = [4,11,20,20,3,15,20,11]
more_nums = [3,15,20,11,4,11,20,20]

print(list(map(lambda x: (x*2)-1, numbers)))
print(list(map(lambda x,y: ((x*2)-1,(y*2)-1), numbers, more_nums)))

[7, 21, 39, 39, 5, 29, 39, 21]
[(7, 5), (21, 29), (39, 39), (39, 21), (5, 7), (29, 21), (39, 39), (21, 39)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [44]:
# syntax for this looks like
# list(filter(function, iterables(s)))
# similar to mask, a 

names = [' ', 'Andy', 'Max', ' ', 'Angela']

def a_names(name):
    # In this function we will check to see if the first letter of the name
    # starts with A
    if name[0].lower() == ' ':
        return False
    else: # If it is not an 'a'
        return True
# Our function returns as a boolean value
# But since this is a list, need to wrap in a list
list(filter(a_names, names))

['Andy', 'Max', 'Angela']

#### Using Lambda's with Filter()

In [42]:
# WE can also use lambdas wit ha filter function
# this is great if you are looking to filter data with a single-use function
# and create another function like we did above
list(filter(lambda name: True if name[0].lower() == 'a' else False, names))

['Andy', 'Angela']

In [23]:
for name in names:
    print(name, name[0], name[1:])

Bob B ob
Andy A ndy
Max M ax
Evan E van
Angela A ngela


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [32]:
import statistics
import numpy as np

# using statistic modules to gather the mean
print(statistics.mean(numbers))

#using numphy method
print(np.mean(numbers))

# Numbers 1: Using vanilla python
# Use vanilla python if possible, just because libraries can be alot
print(list(filter(lambda x: True if x>= sum(numbers)/ len(numbers) else False, numbers)))

# Number 2: Using the numpy library
print(list(filter(lambda x: True if x >= np.mean(numbers) else False, numbers)))

# Number 3: Using the statistics module
print(list(filter(lambda x: True if x>= statistics.mean(numbers) else False, numbers)))


13
13.0
[20, 20, 15, 20]
[20, 20, 15, 20]
[20, 20, 15, 20]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Show a way to break the program by creating an infinite loop

In [33]:
# Start off with defining a function!
def add_nums_breakable(num):
    #without a base case, we will go and crash. VSCode has built in safety
    if num>-20000:
        return num + add_nums_breakable(num-1)
    else:
add_nums_breakable(5)

: 

: 

#### Implementing a Base Case
A basecase is what causes the function to break out of the infinite loop!
It evaluates the value of the number/string at the moment it recurrs to stop it form advancing.

These need to be implemented at the BEGINNING of your function, not after the recursive call.

In [3]:
def add_nums(num):
    #Im plement base case
    if num <= 1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums{num} = {num} + add_nums({num-1})')
    return num + add_nums(num-1)

add_nums(5)

add_nums5 = 5 + add_nums(4)
add_nums4 = 4 + add_nums(3)
add_nums3 = 3 + add_nums(2)
add_nums2 = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [4]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    #Base case to begin with!
    if num <= 1:
        return 1
    else:
        print(f'factorial({num}) = {num} * factorial({num - 1})')
        return num * factorial(num-1)
print(factorial(5))

factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)
120


#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [32]:
def sub_nums(num1, num2):
    #Im plement base case
    if num2 == 0:
        return num1
    else:
        print(num1, num2)
        return subtract_numbers(num1-1, num2 -1)
    return num
sub_nums(25,5)
        




25 5
24 - 4 = 20
20 - 3 = 17
17 - 2 = 15
15 - 1 = 14


14

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [36]:
# Vanilla python generator
#print(type(range(0,10)))

def my_range(stop, start = 0, step = 2):
    while start < stop:
        yield start # Yield means that is going to be a generator. It will not return anything when the function is run
        start += step
[i for i in my_range(10)]


[0, 2, 4, 6, 8]

#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [39]:
# Generators package up the values and yield them as opposed to returning them.
# then we can iterate through that generator object to return each value package
# inside of it
def squared_nums(num):
    while num > 0:
        yield num **2
        num -= 1
[i for i in squared_nums(5)]

[25, 16, 9, 4, 1]

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [63]:
# Use the .filter() 
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
list(filter(str.strip, places))


['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [68]:
# use lambda function
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda name: name.split()[-1].lower())
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]v
`

In [69]:
# F = (9/5)*C + 32
#using the map function with with lambda
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# Using map() with lambda to apply conversion
places_F = list(map(lambda place: (place[0], ((9/5)*place[1]) + 32), places))

print(places_F)



[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`